### ISSUE TITLE
Issues related to training text recognition models based on custom datasets using mmocr-1.x version code.

### ISSUE THREAD

**Issue by htaoruan** [open-mmlab/mmocr#1444]
First I added mydata.py
`my_data_root = '/mmocr/mydata/'

mydata_train = dict(
     type='RecogTextDataset',
     data_root=my_data_root,
     ann_file='train.txt',
     data_prefix=dict(img_path='imgs'),
     parser_cfg=dict(
         type='LineStrParser',
         keys=['filename', 'text'],
         keys_idx=[0, 1]),
     pipeline=[])
     
mydata_test = dict(
     type='RecogTextDataset',
     data_root=my_data_root
     ann_file='val.txt',
     data_prefix=dict(img_path='imgs'),
     parser_cfg=dict(
         type='LineStrParser',
         keys=['filename', 'text'],
         keys_idx=[0, 1]),
     pipeline=[])
`
Then make changes under crnn_mini-vgg_5e_toy.py
`# training schedule for 1x
_base_ = [
    '../_base_/default_runtime.py',
    '../_base_/datasets/my_data.py',
    '../_base_/schedules/schedule_adadelta_5e.py',
    '_base_crnn_mini-vgg.py',
]

# dataset settings
train_list = [_base_.mydata_train]
test_list = [_base_.mydata_test]

default_hooks = dict(logger=dict(type='LoggerHook', interval=50), )

train_dataloader = dict(
    batch_size=64,
    num_workers=8,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='RecogTextDataset',
        datasets=train_list,
        pipeline=_base_.train_pipeline))
val_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='RecogTextDataset',
        datasets=test_list,
        pipeline=_base_.test_pipeline))
test_dataloader = val_dataloader

model = dict(decoder=dict(dictionary=dict(with_unknown=True)))

val_evaluator = dict(dataset_prefixes=['Toy'])
test_evaluator = val_evaluator`

Then I get the following error：
TypeError: class `EpochBasedTrainLoop` in mmengine/runner/loops.py: class `RecogTextDataset` in mmocr/datasets/recog_text_dataset.py: __init__() got an unexpected keyword argument 'datasets'

I would like to ask about training on my own dataset, how to modify the configuration file and whether my changes are correct?

**Comment by xinke-wang** [comment]
As shown in the error report, the `RecogTextDataset` does not accept the `datasets` argument.

If you want to use the `train_list` and `test_list`, you have to use `type='ConcatDataset'`.

see the crnn config as example
https://github.com/open-mmlab/mmocr/blob/dev-1.x/configs/textrecog/crnn/crnn_mini-vgg_5e_mj.py#L29-#L32

**Comment by htaoruan** [comment]
Thank you, after I modified it as you suggested, the following error：
ValueError: class `EpochBasedTrainLoop` in mmengine/runner/loops.py: class `ConcatDataset` in mmocr/datasets/dataset_wrapper.py: class `RecogTextDataset` in mmocr/datasets/recog_text_dataset.py: key index: 1 out of range: ['']

**Comment by xinke-wang** [comment]
The most likely cause of this error is a discrepancy between the number of annotation fields and the number of parser fields.
In your configuration, the `parser` contains **2** keys, i.e., `['filename', 'text']`. Hence the txt annotation should be formatted as follows:
```text
img_1.jpg hello
img_2.jpg world
```
So, please double check your annotation files if there are **Any** lines violates the above format like:
```text
img_1.jpg
hello
```

In addition, `RecogTextDataset` is a deprecated dataset format, which will be removed in the future. We strongly recommend you to convert your datasets to MMOCR 1.x format to avoid such issues. Please refer to [Dataset Migration](https://mmocr.readthedocs.io/en/dev-1.x/migration/dataset.html) for more details.

**Comment by htaoruan** [comment]
OK Also, I would like to ask you a question. what these two lines of code mean？ whether it need to be modified when training on a custom dataset.
`val_evaluator = dict(dataset_prefixes=['Toy'])
test_evaluator = val_evaluator`

**Comment by htaoruan** [comment]
OK Also, I would like to ask you a question. what these two lines of code mean？ whether it need to be modified when training on a custom dataset.
`val_evaluator = dict(dataset_prefixes=['Toy'])
test_evaluator = val_evaluator`

**Comment by xinke-wang** [comment]
For the first line

```
val_evaluator = dict(dataset_prefixes=['Toy'])
```
At evaluation stage, the prefixes play as flags to indicate which dataset does the reported accuracy belongs to, it is actually a list of dataset names that match to the `datasets` list. For example,

https://github.com/open-mmlab/mmocr/blob/f8ccf746c8a9a54403e6c047e0a2418c175532f0/configs/textrecog/crnn/crnn_mini-vgg_5e_mj.py#L17-L20

The above config uses 6 datasets for evaluation, so we define their names in the `val_evaluator`. (It's noteworthy that the order of names in `dataset_prefixes` should be exactly match to the `test_list`)

https://github.com/open-mmlab/mmocr/blob/f8ccf746c8a9a54403e6c047e0a2418c175532f0/configs/textrecog/crnn/crnn_mini-vgg_5e_mj.py#L45-L46

By setting as the above configs, you will see these prefixes in the training logs as follows:

```text
CUTE80/recog/word_acc: 0.1007  CUTE80/recog/word_acc_ignore_case: 0.5417  CUTE80/recog/word_acc_ignore_case_symbol: 0.5417
IIIT5K/recog/word_acc: 0.0490  IIIT5K/recog/word_acc_ignore_case: 0.7803  IIIT5K/recog/word_acc_ignore_case_symbol: 0.7803
SVT/recog/word_acc: 0.7589  SVT/recog/word_acc_ignore_case: 0.7589  SVT/recog/word_acc_ignore_case_symbol: 0.7589
SVTP/recog/word_acc: 0.0000  SVTP/recog/word_acc_ignore_case: 0.5566  SVTP/recog/word_acc_ignore_case_symbol: 0.5566
IC13/recog/word_acc: 0.2601  IC13/recog/word_acc_ignore_case: 0.8542  IC13/recog/word_acc_ignore_case_symbol: 0.8542
IC15/recog/word_acc: 0.0669  IC15/recog/word_acc_ignore_case: 0.4834  IC15/recog/word_acc_ignore_case_symbol: 0.5012  
```

In conclusion, in your case, you may change the `Toy` to the name of your own dataset or any alias you like, and then you will see it in your training log.

For the second line,

```
test_evaluator = val_evaluator
```

It just represents that use the same `test_evaluator` as `val_evaluator`. This is because most of the OCR datasets only have `train` and `test` splits, therefore, we use the same dataset for test/val. It is not necessary to modify this setting unless you want to use different settings/datasets for val and test.

**Comment by gaotongxiao** [comment]
You might find more info at 
https://mmocr.readthedocs.io/en/dev-1.x/user_guides/config.html#evaluation-configuration
and 
https://mmocr.readthedocs.io/en/dev-1.x/basic_concepts/evaluation.html

**Comment by htaoruan** [comment]
Sorry, the network problem caused the reply to send multiple times, thank you very much for your reply

**Comment by htaoruan** [comment]
Hi，In the new version, if I want to save the best weight for 0_word_acc_ignore_case_symbol, where should I change it?

**Comment by xinke-wang** [comment]
Hi, to save the best model in 1.x version, you have to modify the [`CheckpointHook`](https://mmocr.readthedocs.io/en/dev-1.x/user_guides/config.html#hook-configuration).

https://github.com/open-mmlab/mmocr/blob/e467f283d29ee1a8e263dc479c18bfd84c411bdf/configs/textrecog/_base_/default_runtime.py#L13

Assume that your dataset name is `Toy` (which you have defined in `val_evaluator = dict(dataset_prefixes=['Toy'])`), then, the checkpoint hook should be modified as follows:

```python
CheckpointHook(interval=10, by_epoch=True, save_best='Toy/recog/word_acc_ignore_case_symbol', rule='greater')
```

The metric name used here is exactly the same as the one shown in your training logs.

For more information and examples of the CheckpointHook settings, please refer to the [API Doc](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.hooks.CheckpointHook.html#checkpointhook).

**Comment by htaoruan** [comment]
Thank  you very much



---- Replied Message ----
| From | Xinyu ***@***.***> |
| Date | 10/12/2022 17:03 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [open-mmlab/mmocr] Issues related to training text recognition models based on custom datasets using mmocr-1.x version code. (Issue #1444) |

Hi, to save the best model in 1.x version, you have to modify the CheckpointHook.

https://github.com/open-mmlab/mmocr/blob/e467f283d29ee1a8e263dc479c18bfd84c411bdf/configs/textrecog/_base_/default_runtime.py#L13

Assume that your dataset name is Toy (which you have defined in val_evaluator = dict(dataset_prefixes=['Toy'])), then, the checkpoint hook should be modified as follows:

CheckpointHook(interval=10, by_epoch=True, save_best='Toy/recog/word_acc_ignore_case_symbol', rule='greater')

The metric name used here is exactly the same as the one shown in your training logs.

For more information and examples of the CheckpointHook settings, please refer to the API Doc.
